### PR TITLE
fix: Add support for ocischema.DeserializedImageIndex manifest type

### PR DESCRIPTION
**Pull request by jessebye**
I botched the commit history in #617 (rebasing is hard :rage1:) and so started a fresh branch/PR. @jannfis I hope this addresses your feedback about code duplication - I refactored most of the common code into a helper function.

### CODE REVIEW

**Comment by jannfis** [comment]
Thank you, @jessebye. How would you feel about adding some unit testing here? I know it's much of an ask, because there basically are none for the client right now.

File: pkg/registry/client.go [hunk 1]
```diff
@@ -387,8 +349,74 @@ func (client *registryClient) TagMetadata(manifest distribution.Manifest, opts *
...
 	default:
-		return nil, fmt.Errorf("invalid manifest type")
+		return nil, fmt.Errorf("invalid manifest type %T", manifest)
+	}
+}
+
+// TagInfoFromReferences is a helper method to retrieve metadata for a given
+// list of references. It will return the most recent pushed manifest from the
+// list of references.
+func TagInfoFromReferences(client *registryClient, opts *options.ManifestOptions, logCtx *log.LogContext, ti *tag.TagInfo, references []distribution.Descriptor) (*tag.TagInfo, error) {
```
Comment (jannfis) [inline pkg/registry/client.go:359]: Since this is merely an internal helper, I think it could be better prototyped as:

```go
func (client *registryClient) tagInfoFromReferences(...
```

to make it a method of `registryClient`  and make it non-exported.

**Review by jannfis — APPROVED** [review]
LGTM

**Comment by jannfis** [comment]
@jessebye No worries, thanks a lot for your contribution! We'll figure it out eventually and add unit testing along the way :)
